# Pandas1 - Objects

**Import pandas package.**

In [1]:
import pandas as pd

## (1) Series

**(1.1) Create a pandas series using the following list. Print out the series.**

In [2]:
profit_list = [40.4, 63.9, 17.9, 21.3, 51.3, 48.5]

In [3]:
profit_series = pd.Series(profit_list)
print(profit_series)

0    40.4
1    63.9
2    17.9
3    21.3
4    51.3
5    48.5
dtype: float64


**(1.2) Create a pandas series using the list in (1.1) and use the months 'Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun' as index. Print out the series.**

In [4]:
profit_series = pd.Series(profit_list, index = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun'])
print(profit_series)

Jan    40.4
Feb    63.9
Mar    17.9
Apr    21.3
May    51.3
Jun    48.5
dtype: float64


## (2) DataFrame

**(2.1) Create a dictionary containing three key-value pairs. Use the keys "company_name", "profit" and "assets" and specify the following list as the corresponding values.**

In [5]:
company_list = ['JPMorgan Chase','Apple','Bank of America','Amazon','Microsoft']
profit_list = [40.4, 63.9, 17.9, 21.3, 51.3]
assets_list = [3689.3, 354.1, 2832.2, 321.2, 304.1]

In [6]:
data = {"company_name":company_list, "profit":profit_list, "assets": assets_list}
print(data)

{'company_name': ['JPMorgan Chase', 'Apple', 'Bank of America', 'Amazon', 'Microsoft'], 'profit': [40.4, 63.9, 17.9, 21.3, 51.3], 'assets': [3689.3, 354.1, 2832.2, 321.2, 304.1]}


**(2.2) Create a dataframe from the dictionary you obtained in (2.1). Display the dataframe.**

In [7]:
company_df = pd.DataFrame(data)
display(company_df)

company_name  profit  assets
0   JPMorgan Chase    40.4  3689.3
1            Apple    63.9   354.1
2  Bank of America    17.9  2832.2
3           Amazon    21.3   321.2
4        Microsoft    51.3   304.1

**(2.3) Using the dataframe obtained in (2.2), print the number of rows and columns.**

In [8]:
company_df.shape

(5, 3)

**(2.4) What is the average profit of these five companies?**

In [9]:
company_df["profit"].mean()

38.96

**(2.5) Using the dataframe obtained in (2.2), arrange the data in alphabetical order based on the "company_name" column.**

In [10]:
company_df.sort_values(by = "company_name")

company_name  profit  assets
3           Amazon    21.3   321.2
1            Apple    63.9   354.1
2  Bank of America    17.9  2832.2
0   JPMorgan Chase    40.4  3689.3
4        Microsoft    51.3   304.1

## (3) Read data from file 

All the .csv files used in this exercise can be found under the folder **Resourses/Part2: Data Extraction and Visualizaion/Dataset/** on itslearning

### (3.1) Melbourne housing data

**This file <code>melbourne.csv</code> records property sales prices in Melbourne. Descriptions of some important variables are listed below.**

- **Suburb**: Suburb
- **Rooms**: Number of rooms
- **Type**: type of real estate including: 
    - br - bedroom(s)
    - h - house,cottage,villa, semi,terrace
    - u - unit, duplex
    - t - townhouse
    - dev site - development site;
    - o res - other residential.
- **Price**: Price in Australian dollars
- **Distance**: Distance from Central Business Development (in km)
- **Postcode**: Postcode
- **Bathroom**: Number of bathrooms
- **Region**: General Region (West, North West, North, North east …etc)


**(3.1.1) Read the csv file <code>melbourne.csv</code> as a pandas dataframe. Display the first five rows.** 

In [25]:
house_df = pd.read_csv("../dataset/melbourne.csv")
house_df.head(5)

Suburb           Address  Rooms Type      Price Method SellerG  \
0  Abbotsford      85 Turner St      2    h  1480000.0      S  Biggin   
1  Abbotsford   25 Bloomburg St      2    h  1035000.0      S  Biggin   
2  Abbotsford      5 Charles St      3    h  1465000.0     SP  Biggin   
3  Abbotsford  40 Federation La      3    h   850000.0     PI  Biggin   
4  Abbotsford       55a Park St      4    h  1600000.0     VB  Nelson   

        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  3/12/2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
1  4/02/2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
2  4/03/2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   
3  4/03/2017       2.5    3067.0  ...       2.0  1.0      94.0           NaN   
4  4/06/2016       2.5    3067.0  ...       1.0  2.0     120.0         142.0   

   YearBuilt  CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN        Yarra  -37.7996    144.9984  Northern Metropolitan   
1     1900.0        Yarra  -37.8079    144.9934  Northern Metropolitan   
2     1900.0        Yarra  -37.8093    144.9944  Northern Metropolitan   
3        NaN        Yarra  -37.7969    144.9969  Northern Metropolitan   
4     2014.0        Yarra  -37.8072    144.9941  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  

[5 rows x 21 columns]

**(3.1.2) Using the data frame obtained in (3.1.1), print out the number of rows and columns in the dataframe in the following format.**<br>
Expected result:
<code>
Number of observations: 13580
Number of variables: 21
</code>    

In [26]:
print("Number of observations:", house_df.shape[0])
print("Number of variables:", house_df.shape[1])

Number of observations: 13580
Number of variables: 21


**(3.1.3) Show a summary of the dataframe, including column names and their data types.** 

In [13]:
house_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Suburb         13580 non-null  object 
 1   Address        13580 non-null  object 
 2   Rooms          13580 non-null  int64  
 3   Type           13580 non-null  object 
 4   Price          13580 non-null  float64
 5   Method         13580 non-null  object 
 6   SellerG        13580 non-null  object 
 7   Date           13580 non-null  object 
 8   Distance       13580 non-null  float64
 9   Postcode       13580 non-null  object 
 10  Bedroom2       13580 non-null  float64
 11  Bathroom       13580 non-null  float64
 12  Car            13518 non-null  float64
 13  Landsize       13580 non-null  float64
 14  BuildingArea   7130 non-null   float64
 15  YearBuilt      8205 non-null   float64
 16  CouncilArea    12211 non-null  object 
 17  Lattitude      13580 non-null  float64
 18  Longti

**(3.1.4) Show the descriptive statistics for all numerical columns and round the result to one decimal place.**

In [14]:
house_df.describe().round(1)

Rooms      Price  Distance  Bedroom2  Bathroom      Car  Landsize  \
count  13580.0    13580.0   13580.0   13580.0   13580.0  13518.0   13580.0   
mean       2.9  1075684.1      10.1       2.9       1.5      1.6     558.4   
std        1.0   639310.7       5.9       1.0       0.7      1.0    3990.7   
min        1.0    85000.0       0.0       0.0       0.0      0.0       0.0   
25%        2.0   650000.0       6.1       2.0       1.0      1.0     177.0   
50%        3.0   903000.0       9.2       3.0       1.0      2.0     440.0   
75%        3.0  1330000.0      13.0       3.0       2.0      2.0     651.0   
max       10.0  9000000.0      48.1      20.0       8.0     10.0  433014.0   

       BuildingArea  YearBuilt  Lattitude  Longtitude  Propertycount  
count        7130.0     8205.0    13580.0     13580.0        13580.0  
mean          152.0     1964.7      -37.8       145.0         7454.4  
std           541.0       37.3        0.1         0.1         4378.6  
min             0.0     1196.0      -38.2       144.4          249.0  
25%            93.0     1940.0      -37.9       144.9         4380.0  
50%           126.0     1970.0      -37.8       145.0         6555.0  
75%           174.0     1999.0      -37.8       145.1        10331.0  
max         44515.0     2018.0      -37.4       145.5        21650.0

**(3.1.5) Use the result obtained in (3.1.4) to answer the following questions.** (Add your answers to the markdown cell.) 

- What is the average selling price of the properties? 1075684.1
- Among these properties, what is the highest number of rooms in a single property? 10
- What is the average number of bathrooms in these properties? 1.5
- Among these properties, in which year was the oldest property built? 1196

**(3.1.6) In which three suburbs are most properties located?**

In [15]:
house_df["Suburb"].value_counts().head(3)

Reservoir         359
Richmond          260
Bentleigh East    249
Name: Suburb, dtype: int64

**(3.1.7) Display the data of the top five properties with the highest prices.**<br>
Hint: sorting

In [16]:
house_df.sort_values(by = "Price", ascending = False).head(5)

Suburb         Address  Rooms Type      Price Method  \
12094     Mulgrave     35 Bevis St      3    h  9000000.0     PI   
7692    Canterbury  49 Mangarra Rd      5    h  8000000.0     VB   
9575      Hawthorn    49 Lisson Gr      4    h  7650000.0      S   
3616           Kew     15 Barry St      6    h  6500000.0      S   
12557  Middle Park     136 Page St      5    h  6400000.0      S   

            SellerG        Date  Distance Postcode  ...  Bathroom  Car  \
12094          Hall  29/07/2017      18.8   3170.0  ...       1.0  1.0   
7692      Sotheby's  13/05/2017       9.0   3126.0  ...       5.0  4.0   
9575   Abercromby's  17/06/2017       5.3   3122.0  ...       2.0  4.0   
3616         Jellis  13/08/2016       5.6   3101.0  ...       6.0  3.0   
12557      Marshall   9/09/2017       3.0   3206.0  ...       2.0  1.0   

       Landsize  BuildingArea  YearBuilt  CouncilArea Lattitude  Longtitude  \
12094     744.0         117.0     1960.0       Monash -37.93168   145.16126   
7692     2079.0         464.3     1880.0   Boroondara -37.81790   145.06940   
9575     1690.0         284.0     1863.0   Boroondara -37.82652   145.03052   
3616     1334.0         365.0     1890.0   Boroondara -37.80290   145.02670   
12557     553.0         308.0     1920.0          NaN -37.84908   144.95753   

                       Regionname Propertycount  
12094  South-Eastern Metropolitan        7113.0  
7692        Southern Metropolitan        3265.0  
9575        Southern Metropolitan       11308.0  
3616        Southern Metropolitan       10331.0  
12557       Southern Metropolitan        2019.0  

[5 rows x 21 columns]

### (3.2) The Diabetes dataset

• **Pregnancies**: Number of times pregnant<br>
• **Glucose**: Plasma glucose concentration over 2 hours in an oral glucose tolerance test<br>
• **BloodPressure**: Diastolic blood pressure (mm Hg)<br>
• **SkinThickness**: Triceps skin fold thickness (mm)<br>
• **Insulin**: 2-Hour serum insulin (mu U/ml)<br>
• **BMI**: Body mass index (weight in kg/(height in m)2)<br>
• **DiabetesPedigreeFunction**: Diabetes pedigree function (a function which scores likelihood of
diabetes based on family history)<br>
• **Age**: Age (years)<br>
• **Outcome**: Class variable (0 if non-diabetic, 1 if diabetic)<br>

**(3.2.1) Import the csv file <code>diabetes.csv</code> and specify the data type for the 'Outcome' column as 'object'. Display the first five rows.** <br>
Hint:
```Python
        pd.read_csv("file_name", dtype = {"column_name": object})
```

In [17]:
diabetes_df = pd.read_csv("../dataset/diabetes.csv", dtype = {"Outcome": object})
diabetes_df.head(5)

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age Outcome  
0                     0.627   50       1  
1                     0.351   31       0  
2                     0.672   32       1  
3                     0.167   21       0  
4                     2.288   33       1

**(3.2.2) Show the descriptive statistics for all numerical variables, rounding the result to the nearest whole number. Then, answer the following questions.**<br>

In [18]:
diabetes_df.describe().round()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin    BMI  \
count        768.0    768.0          768.0          768.0    768.0  768.0   
mean           4.0    121.0           69.0           21.0     80.0   32.0   
std            3.0     32.0           19.0           16.0    115.0    8.0   
min            0.0      0.0            0.0            0.0      0.0    0.0   
25%            1.0     99.0           62.0            0.0      0.0   27.0   
50%            3.0    117.0           72.0           23.0     30.0   32.0   
75%            6.0    140.0           80.0           32.0    127.0   37.0   
max           17.0    199.0          122.0           99.0    846.0   67.0   

       DiabetesPedigreeFunction    Age  
count                     768.0  768.0  
mean                        0.0   33.0  
std                         0.0   12.0  
min                         0.0   21.0  
25%                         0.0   24.0  
50%                         0.0   29.0  
75%                         1.0   41.0  
max                         2.0   81.0

- What is the average age? 33
- What is the maximum BMI? 67
- What is the average BloodPressure? 69

**(3.2.3) Display data for the top 15 people with the highest BMI. Store the data as a new dataframe.**

In [19]:
bmi_top15 = diabetes_df.sort_values(by = "BMI", ascending = False).head(15)
display(bmi_top15)

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
177            0      129            110             46      130  67.1   
445            0      180             78             63       14  59.4   
673            3      123            100             35      240  57.3   
125            1       88             30             42       99  55.0   
120            0      162             76             56      100  53.2   
303            5      115             98              0        0  52.9   
247            0      165             90             33      680  52.3   
193           11      135              0              0        0  52.3   
155            7      152             88             44        0  50.0   
99             1      122             90             51      220  49.7   
681            0      162             76             36        0  49.6   
746            1      147             94             41        0  49.3   
84             5      137            108              0        0  48.8   
378            4      156             75              0        0  48.3   
335            0      165             76             43      255  47.9   

     DiabetesPedigreeFunction  Age Outcome  
177                     0.319   26       1  
445                     2.420   25       1  
673                     0.880   22       0  
125                     0.496   26       1  
120                     0.759   25       1  
303                     0.209   28       1  
247                     0.427   23       0  
193                     0.578   40       1  
155                     0.337   36       1  
99                      0.325   31       1  
681                     0.364   26       1  
746                     0.358   27       1  
84                      0.227   37       1  
378                     0.238   32       1  
335                     0.259   26       0

**(3.2.4) Using the dataframe obtained in (3.2.3), count the number of people with and without diabetes.**

In [20]:
bmi_top15["Outcome"].value_counts()

1    12
0     3
Name: Outcome, dtype: int64

### (3.3) Country code

**(3.3.1) Read "code.csv" as a dataframe aisplay the first 10 rows.**

In [21]:
code_df = pd.read_csv("../dataset/code.csv")
code_df.head(10)

country_code              country    region
0          AFG          Afghanistan      Asia
1          ALA        Åland Islands    Europe
2          ALB              Albania    Europe
3          DZA              Algeria    Africa
4          ASM       American Samoa   Oceania
5          AND              Andorra    Europe
6          AGO               Angola    Africa
7          AIA             Anguilla  Americas
8          ATA           Antarctica       NaN
9          ATG  Antigua and Barbuda  Americas

**(3.3.2) Extract the  "country_code" column and "country" column as two lists.**<br> Hint: <code>df["column_name"].to_list()</code>

In [22]:
code_list = code_df["country_code"].to_list()
country_list = code_df["country"].to_list()

**(3.3.3) Using the two lists obtained in (3.3.2), create a dictionary with "country code" as the key and "country" as the value.**<br>
Example: <br>
<code>{"AFG": "Afghanistan", "ALA": "Åland Islands",..., "ZWE": "Zimbabwe"}</code>

In [23]:
country_dict = dict(zip(code_list, country_list))

**(3.3.4) Write a program that asks the user to enter a country code, and the program should print out the corresponding country name.**<br>
Example-1:<br>
Enter a country code: <code>AGO</code>.<br>
The country code "AGO" is Angola.

Example-2:<br>
Enter a country code: <code>NOR</code>.<br>
The country code "NOR" is Norway.

In [24]:
code = input("Enter a country code: ")

print(f'The country code "{code}" is {country_dict[code]}.')

Enter a country code: AGO
The country code "AGO" is Angola.
